### ISSUE TITLE
Copying pre-existing sqlite databases

### ISSUE THREAD

**Issue by danbenn** [capacitor-community/sqlite#61]
Is copying pre-existing sqlite databases: 

A: a feature that should live in a separate package, or,
B: something that would be welcome in a pull request for this package?

There is this existing package: https://github.com/an-rahulpandey/cordova-plugin-dbcopy

But it's difficult to use with capacitor. 

Happy to look into writing this functionality, just wanted feedback on where that should live.

@jepiqueau

**Comment by jepiqueau** [comment]
@danbenn You have to be a little more specific on what you want the copy to be. Where is residing the database that you want to copy ? Is there in the same folder? do you want to copy it to a new name? what about security of one copying your database ?
Currently, if the database is residing on a server you can use the importFromJson meaning that you have to do the function which is reading your database from the server and convert it in the desired json format as you will do for synchronization.
If the database is residing on the platform database folder and you want to copy it to a new name database this can be easily added with the following process for the refactor version
```ts
// create a connection for the database you want to copy ie "NoEncryption"
 const db = await sqlite.createConnection("NoEncryption");
// open the database
const res = db.open();
if(res.result) const ret = db.copy("NewCopyName");
 ```
this is easy to add as all the methods are already available in the plugin, just have to add the copy method to the interface
if something else you have to specify what you have in mind.
Thanks for your using the plugin and for your constructive remarks to enhance the plugin

**Comment by danbenn** [comment]
@jepiqueau thanks for your thoughtful response.

I have a pre-populated sqlite database that lives in the assets folder.

The idea is that I could copy it from the assets to a folder where the database can be read.

Alternatively, I could download the database from a server and load it that way, just seems like unnecessary use of server resources if the user can get the database right from within the bundle.

Thoughts?

**Comment by jepiqueau** [comment]
@danbenn, Thanks for clarifying, from the asset folder of your app it should be feasible. Are you using the `refactor` release of the plugin? Is it something urgent as i am currently busy on developing the Vue app starter for the refactor? I will have a look to implement it in the next release of the refactor, it will be a new method like `copyFromAsset() 
```ts
const res = await sqlite.copyFromAsset();
```
the process flow will be
```ts
// > look for all databases in the asset folder
// > Loop on each database
// >> add the suffix SQLite.db to the database name
// >> copy the database from the asset folder to the database device folder
```

**Comment by jepiqueau** [comment]
@danbenn This is solved in the release 2.9.0-beta.2 of the plugin. I did not updated yet the ionic/Angular and the Ionic/React app starters but i did it for the Ionic/Vue. so that could give you and idea. The database are stored in public/assets/databases folder of the app. All files with the `.db` extension  in the public/assets/databases are copied into the database folder of each device when you execute the  following method
```ts
let res: any = await sqlite.copyFromAssets()
```
You will notice in the Ionic/Vue app that the database could have or not the `SQLite.db`  this suffix will be added during the copy if not existing

voilà

**Comment by danbenn** [comment]
That was extremely fast, thank you @jepiqueau !! 👏 🎉

**Comment by HarelM** [comment]
I have a follow question related to this issue (somewhat).
I've tried to read the docs but I'm not sure I fully understand them.
My use case is the following:
1. I download a sqlite database from the server.
2. I place it in the cache folder of the app (It's a large database ~ 100 Mb so I need to use cordova-file-transfer plugin to download it and I would like to continue working with the existing database until this operation completes)
3. After the download is complete I close the existing database and move the database to a different folder using capacitor Filesystem API

My question is where to I need to place the file in order for this plugin to read it (relevant for both iOS and Android) in Filesystem API terms? Data folder for Android and Documents for iOS?
I don't want this file to be uploaded to the iCould (i.e. I want it to remain "secret"), do I need to configure this as described in the docs in a different folder? If so, how would I use the Filesystem API to move it to the right location?

Thanks again for all this! I'm in the middle of migration from cordova to capacitor and I'm currently using cordova-sqlite-ext plugin which I'm not sure is working well in capacitor and I prefer a more "capacitor native" solution.

**Comment by jepiqueau** [comment]
@HarelM thanks for using the plugin. There are methods provided for this in the plugin.
After your db has been copied with the cordova plugin in a specific folder you can use
- GetMigratableDbList at the path of the specific folder to make sure that the db is there.
- addSQLiteSuffix to make the copy of  of the db to the right folder and add SQLite suffix to the db 
- deleteOldDatabases to remove the file from the specific folder.

**Comment by jepiqueau** [comment]
@HarelM you can have a look at the MigratingCordovaDatabases doc. If you do not want to have your db visible you can define in the capacitor-config.ts the location of your databases see https://github.com/jepiqueau/angular-sqlite-app-starter . Hope this clarify.

**Comment by HarelM** [comment]
Thanks @jepiqueau, super useful!
So basically I need to use the addsqlitesuffix and give it the cache folder, right? This means I don't need to worry about copying the databased my self and I don't need to know where the database actually are? Cool!

**Comment by jepiqueau** [comment]
@HarelM the database will be in the location specify in the capacitor-config.ts otherwise they will be à the default location. Test it and if you find an issue tell me. Have a good migration of your app

**Comment by HarelM** [comment]
I used the method you said - `addSQLiteSuffix` and `deleteOldDatabases` but when I open a connection it tells me that there are no tables.
I'm guessing that I'm doing something wrong in terms of opening the connection as it's probably creating a new database instead of using the existing.
Please let me know what I'm doing wrong...
the filename is for example `something.mbtiles` downloaded to cache folder.
I run 
`await this.sqlite.addSQLiteSuffix("cache", [something.mbtiles]);`
`await this.sqlite.createConnection("something", ...)` (or should I be trying to connect to `something.mbtiles`? although in the logcat it looks like it is using the right database)

**Comment by jepiqueau** [comment]
@HarelM may be it is .mbtiles. could you call it something_mbtiles. In that case you should create the connection as await this.createConnection("something_mbtiles",...)

**Comment by jepiqueau** [comment]
@HarelM forget about what i said something.mbtiles should work 
await this.sqlite.createConnection(something.mbtiles,...)
Use method getTableList to get the liste of tables in the database

**Comment by HarelM** [comment]
Nothing is working for me :-(( I'm sure I'm missing something obvious...
I've tried to save the databases as .db files and remove all the mess around .mbtiles stuff but it doesn't seem to help.
I still see some databases in the database list.
I'll try to clean the app storage stuff and see.
Is it possible that replacing an existing database doesn't really replace it?

**Comment by jepiqueau** [comment]
@HarelM well i never test with a file coming from the cache usually the path that i tested where physical pathes on the device itself.

**Comment by jepiqueau** [comment]
@HarelM how do you download the database from the server http request? Have you made this process when you were using the cordova sqlite plugin. Can you share your app or a repro app i may have a look.

**Comment by HarelM** [comment]
I'm using cordiva plugin file transfer.
The repo is here:
https://github.com/IsraelHikingMap/Site/tree/main/IsraelHiking.Web/src/application/services
Checkout the file.service and database.service in main and capacitor-migration branches.
The migration is still a work in progress...

**Comment by jepiqueau** [comment]
@HarelM i will have a look when i am back home and will tell you if i fond something

**Comment by HarelM** [comment]
OK, after cleaning up everything and making sure I don't open a connection before checking that the database exists I get the following error:
`Error: getDatabaseList: No databases available` so this means that there's something wrong with the copying of the databases.
HMM... come to think about it, I think I experienced this issue in the past. the problem was that I was trying to copy the files instead of moving them, which took a long time.
Is there a way to move the files without just copy them? i.e. "merge" the suffix and deleteOldDatabases commands so that I don't need to wait for the FS to finish copying and do it in a more "transaction" kind of way - close the DB, move the file to overwrite exiting file and then open the DB?
It might not be the case, I'll check with the phone in a few minutes to see if the databases are available...

**Comment by HarelM** [comment]
I think I found the problem in the code, please let me know if you would like me to open a new issue in order to track this.
The problem is with the following method:
https://github.com/capacitor-community/sqlite/blob/master/android/src/main/java/com/getcapacitor/community/database/sqlite/SQLite/UtilsMigrate.java#L79
It doesn't know what to do with the `cache` folder and so it returns a different folder - the `files` folder, where there are no databases, and this propagates forward, I believe...
I'm honestly happy to see that I'm using the code of this library correctly and that the problem is here and not with my code :-) (I don't mean to gloat, don't get me wrong :-))

**Comment by HarelM** [comment]
BTW, I needed to add the following to build.gradle after adding this plugin:
```
packagingOptions {
        exclude 'build-data.properties'
}
```
Is there anything that can be done in order to avoid this? I'm guessing that this plugin and another one are using the same file...?

**Comment by jepiqueau** [comment]
@HarelM would be better to open a new issue with à title somethink like this cannot migrate à Cordova database from cache. Then make a link to this one to not loose the history. I have to look how files are stored in the cache. Have you try on iOS? I never had to set  exclude build.data.properties. i am busy on private things for the next two days i will start looking à this after.

**Comment by jepiqueau** [comment]
@HarelM i found that link to move à file from cache to internal  http://www.java2s.com/Code/Android/File/Moveafilestoredinthecachetotheinternalstorageofthespecifiedcontext.htm may be if you have time you should be looking at it and see how to implement it in utilsMigrate by adding a case cache

**Comment by binaryking979** [comment]
@HarelM Hello, I have some issues using capacitor plugin.
I am going to download sqlite database file from server in android device.
And then I am going to connect this database.
But I cannot connect it.
I know you had experience for this problem.
Can you help me?

**Comment by HarelM** [comment]
@jepiqueau is maintaining this package. 
We have moved to use a different local storage mechanism instead of sqlite to avoid this issue (and other improvements).
Thanks for the support @jepiqueau!
